packages/flutter/lib/src/widgets/sensitive_content.dart

```diff
@@ -0,0 +1,373 @@
+  /// that this setting tracks.
+  ContentSensitivity? get contentSensitivityBasedOnWidgetCounts {
+    if (_sensitiveWidgetCount > 0) {
+      return ContentSensitivity.sensitive;
+    }
+    if (_autoSensitiveWidgetCount > 0) {
+      return ContentSensitivity.autoSensitive;
+    }
+    if (_notSensitiveWigetCount > 0) {
+      return ContentSensitivity.notSensitive;
+    }
+    return null;
+  }
+}
+
+/// Host of the current content sensitivity for the widget tree that contains
+/// some number [SensitiveContent] widgets.
+@visibleForTesting
+class SensitiveContentHost {
+  SensitiveContentHost._();
+
+  bool? _contentSenstivityIsSupported;
+  late final _ContentSensitivitySetting _contentSensitivitySetting = _ContentSensitivitySetting();
+  ContentSensitivity? _fallbackContentSensitivitySetting;
+
+  final SensitiveContentService _sensitiveContentService = SensitiveContentService();
+
+  /// [SensitiveContentHost] instance for the widget tree.
+  @visibleForTesting
+  static final SensitiveContentHost instance = SensitiveContentHost._();
+
+  /// Returns the current content sensitivity as tracked by [_contentSensitivitySetting].
+  @visibleForTesting
+  ContentSensitivity? get calculatedContentSensitivity =>
+      _contentSensitivitySetting.contentSensitivityBasedOnWidgetCounts;
+
+  /// Registers a [SensitiveContent] widget that will help determine the
+  /// [ContentSensitivity] for the widget tree.
+  static Future<void> register(ContentSensitivity desiredSensitivity) {
+    return instance._register(desiredSensitivity);
+  }
+
+  Future<void> _register(ContentSensitivity desiredSensitivity) async {
+    _contentSenstivityIsSupported ??= await _sensitiveContentService.isSupported();
+    if (!_contentSenstivityIsSupported!) {
+      // Setting content sensitivity is not supported on this device.
+      return;
+    }
+    // When the first `SensitiveContent` widget is registered, determine the content sensitivity
+    // we should fallback to if/when no `SensitiveContent` widgets remain in the tree.
+    // For Android API 35, this will be auto sensitive if it is otherwise unset by the developer.
+    if (_fallbackContentSensitivitySetting == null) {
+      try {
+        _fallbackContentSensitivitySetting = await _sensitiveContentService.getContentSensitivity();
+      } on UnsupportedError catch (e) {
+        // Unknown ContentSensitivity detected; fallback to not sensitive mode since we
+        // cannot determine the desired behavior of the current mode and log error to user.
+        _fallbackContentSensitivitySetting = ContentSensitivity.notSensitive;
+        FlutterError.reportError(
+          FlutterErrorDetails(
+            exception: FlutterError(
+              'Unknown content sensitivity set in the Android embedding or by default: $e}',
+            ),
+            library: 'widget library',
+            stack: StackTrace.current,
```

Comment:
Nit: Would it be better to get the stacktrace from `catch`? Or is it not passed here?